autobahn/twisted/test/test_protocol.py

```diff
@@ -120,3 +122,111 @@ def test_handshake_fails(self):
+        arg = self.proto.failHandshake.mock_calls[0][1][0]
+        self.assertTrue('not allowed' in arg)
+
+    def test_match_wrong_scheme_origin(self):
+        self.factory.isSecure = False
+        self.proto.data = b"\r\n".join([
+            b'GET /ws HTTP/1.1',
+            b'Host: www.example.com',
+            b'Sec-WebSocket-Version: 13',
+            b'Origin: https://www.example.com',
+            b'Sec-WebSocket-Extensions: permessage-deflate',
+            b'Sec-WebSocket-Key: tXAxWFUqnhi86Ajj7dRY5g==',
+            b'Connection: keep-alive, Upgrade',
+            b'Upgrade: websocket',
+            b'\r\n',  # last string doesn't get a \r\n from join()
+        ])
+        self.proto.consumeData()
+
+        self.assertTrue(self.proto.failHandshake.called, "Handshake should have failed")
+        arg = self.proto.failHandshake.mock_calls[0][1][0]
+        self.assertTrue('not allowed' in arg)
+
+    def test_match_origin_secure_scheme(self):
+        self.factory.isSecure = True
+        self.proto.data = b"\r\n".join([
+            b'GET /ws HTTP/1.1',
+            b'Host: www.example.com',
+            b'Sec-WebSocket-Version: 13',
+            b'Origin: https://www.example.com',
+            b'Sec-WebSocket-Extensions: permessage-deflate',
+            b'Sec-WebSocket-Key: tXAxWFUqnhi86Ajj7dRY5g==',
+            b'Connection: keep-alive, Upgrade',
+            b'Upgrade: websocket',
+            b'\r\n',  # last string doesn't get a \r\n from join()
+        ])
+        self.proto.consumeData()
+
+        self.assertFalse(self.proto.failHandshake.called, "Handshake should have succeeded")
+
+    def test_match_origin_examples(self):
+        """
+        All the example origins from RFC6454 (3.2.1)
+        """
+        # we're just testing the low-level function here...
+        from autobahn.websocket.protocol import _is_same_origin
+        policy = wildcards2patterns(['example.com'])
+
+        self.assertTrue(_is_same_origin('http://example.com/', 80, policy))
+        self.assertTrue(_is_same_origin('http://example.com:80/', 80, policy))
+        self.assertTrue(_is_same_origin('http://example.com/path/file', 80, policy))
+
+    def test_match_origin_counter_examples(self):
+        """
+        All the example 'not-same' origins from RFC6454 (3.2.1)
+        """
+        # we're just testing the low-level function here...
+        from autobahn.websocket.protocol import _is_same_origin
+        policy = wildcards2patterns(['example.com'])
+
+        self.assertFalse(_is_same_origin('http://example.com:8080/', 80, policy))
+        self.assertFalse(_is_same_origin('http://www.example.com/', 80, policy))
+        self.assertFalse(_is_same_origin('https://example.com:80/', 443, policy))
+        self.assertFalse(_is_same_origin('https://example.com/', 80, policy))
+        self.assertFalse(_is_same_origin('http://example.org/', 80, policy))
+        self.assertFalse(_is_same_origin('http://ietf.org/', 80, policy))
```

Comment:
Additional valid URIs, though i'm unsure if these would be possible to spoof into the Origin given any browser:

* `http://example.com./`
* `//example.com/`
* `http://@example.com`